chatApp/chatApp/clientServer.cpp

```diff
@@ -0,0 +1,353 @@
+#pragma comment(lib, "Ws2_32.lib")
+
+TCPserver::TCPserver(std::string ipAddress, int port, MessageReceivedHandler handler)
+	: m_ipAddress(ipAddress), m_port(port), MessageReceived(handler)
+{
+}
+
+TCPserver::TCPserver()
+{
+
+}
+
+//TCPserver::~TCPserver()
+//{
+//	WSACleanup(); 
+//}
+
+bool TCPserver::init()
+{
+	WSADATA wsData;
+
+	int iResult;
+	iResult = WSAStartup(MAKEWORD(2, 2), &wsData);
+
+	return iResult == 0;
+}
+
+SOCKET TCPserver::createSocket()
+{
+	// Create a socket
+	SOCKET listening = socket(AF_INET, SOCK_STREAM, 0);
+	if (listening != INVALID_SOCKET)
+	{
+		SOCKET sock = bindSocket(listening);
+		return sock;
+	}
+	return -1;
+}
+
+SOCKET TCPserver::bindSocket(SOCKET listening)
+{
+	sockaddr_in hint;
+	hint.sin_family = AF_INET;
+	hint.sin_port = htons(m_port);
+	inet_pton(AF_INET, m_ipAddress.c_str(), &hint.sin_addr);
+	// Bind the socket to an IP address and port 
+	int bindOk = bind(listening, (sockaddr*)&hint, sizeof(hint));
+	if (bindOk != SOCKET_ERROR)
+	{
+		int listenOk = listen(listening, SOMAXCONN);
+		if (listenOk == SOCKET_ERROR)
+		{
+			return -1;
+		}
+	}
+	else
+	{
+		return -1;
+	}
+
+	return listening;
+}
+
+SOCKET TCPserver::waitForConnection(SOCKET listening)
+{
+	SOCKET client = accept(listening, NULL, NULL);
+
+	return client;
+}
+
+void TCPserver::run()
+{
+	while (true)
+	{
+		//Create a listening socket
+		SOCKET listening = createSocket();
+		if (listening == INVALID_SOCKET)
+		{
+			cleanup();
+		}
+		//Wait for connection
+		SOCKET client = waitForConnection(listening);
+		if (client != INVALID_SOCKET)
+		{
+			closesocket(listening);
+			//Loop receive/ send
+			recvData(client);
+			closesocket(client);
+		}
+	}
+}
+
+void TCPserver::run_multiple()
+{
+	char buf[MAX_BUFFER_SIZE];
+
+	//Create a listening socket
+	SOCKET listening = createSocket();
+	if (listening == INVALID_SOCKET)
+	{
+		cleanup();
+	}
+	listen(listening, SOMAXCONN);
+
+	fd_set master;
+	FD_ZERO(&master);
+	FD_SET(listening, &master);
+
+	while (true)
+	{
+		fd_set copy = master;
+		int socketCount = select(0, &copy, nullptr, nullptr, nullptr);
+		for (int i = 0; i < socketCount; i++)
+		{
+			SOCKET sock = copy.fd_array[i];
+			if (sock == listening)
+			{
+				// Accept a new connection
+				SOCKET client = accept(listening, nullptr, nullptr);
+				// Add new connection to the list of connected clients
+				FD_SET(client, &master);
+				// Send a welcone message to the connected client
+				TCPpresenter *ptr = new TCPcommandLinePresenter();
+				ptr->welcomeServer(client);
+			}
+			else
+			{
+				recvOrSend_multiple(sock, master, buf, listening);
+			}
+		}
+	}
+}
+
+void TCPserver::sendMsg(int clientSocket, std::string msg)
+{
+	send(clientSocket, msg.c_str(), msg.size() + 1, 0);
+}
+
+void TCPserver::recvData(SOCKET client)
+{
+	char buf[MAX_BUFFER_SIZE];
+	int bytesReceived = 0;
+	do
+	{
+		ZeroMemory(buf, MAX_BUFFER_SIZE);
+		bytesReceived = recv(client, buf, MAX_BUFFER_SIZE, 0);
+		if (bytesReceived > 0)
+		{

```

Comment:
może można by było przerobić tego while i w przypisywać te bytesReceived w while bezpośrednio? Bo teraz jest tak, że i if i while sprawdza Ci ten sam warunek:
https://stackoverflow.com/questions/31659243/c-declaring-and-testing-variable-in-while-condition